src/Features/CSharp/Portable/CodeRefactorings/ConvertToRecord/ConvertToRecordHelpers.cs

```diff
@@ -0,0 +1,914 @@
+                    TryAddEqualizedFieldsForCondition(andOp.RightOperand, successRequirement, currentObject, otherObject, builder),
+            // Exhaustive binary operator for inverted checks via DeMorgan's law
+            // We see an or here, but we're in a context where this being true will return false
+            // for example: return !(expr || expr)
+            // or: if (expr || expr) return false;
+            (false, IBinaryOperation { OperatorKind: BinaryOperatorKind.ConditionalOr } orOp)
+                => TryAddEqualizedFieldsForCondition(orOp.LeftOperand, successRequirement, currentObject, otherObject, builder) &&
+                    TryAddEqualizedFieldsForCondition(orOp.RightOperand, successRequirement, currentObject, otherObject, builder),
+            // we are actually comparing two values that are potentially members,
+            // e.g: return A == other.A;
+            (true, IBinaryOperation
+            {
+                OperatorKind: BinaryOperatorKind.Equals,
+                LeftOperand: IMemberReferenceOperation leftMemberReference,
+                RightOperand: IMemberReferenceOperation rightMemberReference,
+            }) => TryAddFieldFromComparison(leftMemberReference, rightMemberReference, currentObject, otherObject, builder),
+            // we are comparing two potential members, but in a context where if the expression is true, we return false
+            // e.g: return !(A != other.A); 
+            (false, IBinaryOperation
+            {
+                OperatorKind: BinaryOperatorKind.NotEquals,
+                LeftOperand: IMemberReferenceOperation leftMemberReference,
+                RightOperand: IMemberReferenceOperation rightMemberReference,
+            }) => TryAddFieldFromComparison(leftMemberReference, rightMemberReference, currentObject, otherObject, builder),
+            // equals invocation, something like: A.Equals(other.A)
+            (true, IInvocationOperation
+            {
+                TargetMethod.Name: nameof(Equals),
+                Instance: IMemberReferenceOperation invokedOn,
+                Arguments: [IMemberReferenceOperation arg]
+            }) => TryAddFieldFromComparison(invokedOn, arg, currentObject, otherObject, builder),
+            // some other operation, or an incorrect operation (!= when we expect == based on context, etc).
+            // If one of the conditions is just a null check on the "otherObject", then it's valid but doesn't check any members
+            // Otherwise we fail as it has unknown behavior
+            _ => IsNullCheck(condition, successRequirement, otherObject)
+        };
+
+        /// <summary>
+        /// Same as <see cref="TryAddEqualizedFieldsForCondition"/> but we're looking for
+        /// a binding through an "is" pattern first/>
+        /// </summary>
+        /// <returns>the cast parameter symbol if found, null if not</returns>
+        private static bool TryAddEqualizedFieldsForConditionWithoutBinding(
```

Comment:
I admit I don't understand this method name.

Reply:
The idea is we're in a condition (either return or if statement), where we might be checking fields, but we don't have a variable that's bound to the current type yet (i.e. we only have the parameter, which is an `object`). We're looking for that variable to be declared via something of the form `other is C otherC`, where we can now use `otherC` as a variable to check instance fields on.

Reply:
I've changed it to `TryAddEqualizedFieldsForConditionWithoutTypedVariable`, but I'm not sure if it's that much better.